src/lib.rs

```diff
@@ -82,35 +84,87 @@ impl Error for Mcp2210Error {
     fn source(&self) -> Option<&(dyn Error + 'static)> {
         use crate::Mcp2210Error::*;
         match self {
-            Io(err) => Some(err),
+            Hid(err) => Some(err),
             _ => None,
         }
     }
 }
 
-pub type Buffer = [u8; 64];
+const BUFFER_SIZE: usize = 64;
+
+pub type Buffer = [u8; BUFFER_SIZE];
 
 pub const MAX_BIT_RATE: u32 = 12_000_000;
 
 pub struct Mcp2210 {
-    file: File,
+    device: HidDevice,
 }
 
 impl CommandResponse for Mcp2210 {
-    fn command_response(&mut self, cmd: &Buffer, res: &mut Buffer) -> io::Result<()> {
-        self.file.command_response(cmd, res)
+    fn command_response(&mut self, cmd: &Buffer, res: &mut Buffer) -> HidResult<()> {
+        let data_to_write = &[[0x00].to_vec(), cmd.to_vec()].concat();
+        // At this point, length of data_to_write will be BUFFER_SIZE + 1 == 65 or smaller and responses
+        // from the MCP2210 are always BUFFER_SIZE. Therefore, this should only take single reports and
+        // these asserts should be good assumptions.
+        assert_eq!(self.device.write(data_to_write)?, data_to_write.len());
+        assert_eq!(self.device.read(res)?, BUFFER_SIZE);
+        Ok(())
     }
 }
 
 impl Mcp2210 {
-    pub fn open<P: AsRef<Path>>(path: P) -> Result<Mcp2210, Mcp2210Error> {
-        let file = OpenOptions::new()
-            .read(true)
-            .write(true)
-            .open(path)
-            .map_err(Mcp2210Error::Io)?;
-        Ok(Mcp2210 { file })
+    /// Opens a MCP2210 by path
```

Comment:
Actually, now that I think about it, shouldn't the API be just `pub fn new(device: HidDevice) -> Mcp2210`, since we're not doing any actual "opening" and there are no fallible operations :thinking:

Reply:
You're completely right about `new` being used to create a blank instance when the type is a data structure of some kind or maybe some kind of stateful thing with a meaningful empty/initial state. However, in our case I think having a function called `from` has two problems:

- `from` is by convention usually implemented via the `From` / `Into` traits and not as a separate stand-alone function on a type. For example, you mentioned `String::from(..)` and that actually comes from the trait implementation `From<&str> for String`. There's nothing preventing us from having a normal function like that, but I think the naming is confusing since I'd generally expect `Something::from` to involve the `From` / `Into` traits
- we don't really have a data structure but a *wrapper*  over a "raw" HID device. In these kind of cases `new` seems more appropriate and there are several examples of this in the standard library. Adding buffering to a reader by wrapping it with a `BufReader`? `BufReader::new(the_inner_reader)`. Wrapping some data with a mutex? `Mutex::new(the_data)`

In general I view `new` as the "default constructor" for a type, and for wrappers that is usually `pub fn new(the_inner_thing: SomeType)`, assuming instantiating the wrapper always requires the inner thing.